exec-engine.md

```diff
@@ -0,0 +1,146 @@
+# L2 Execution Engine
+
+The rollup mirrors the layer abstraction that the Merge introduces to Ethereum L1:
+- Consensus: *Determine* canonical chain, propagate *blocks*, maintain *system-state*
+- Execution: *Process* canonical chain, propagate *transactions*, maintain *user-state*
+
+The consensus task is implemented by the rollup-node, deriving most functionality from L1.
+The execution task is implemented by the engine, providing an equivalent user environment as L1, with deposits.
+
+This document outlines the modifications, configuration and usage of a L1 execution engine for L2.
+
+
+## Deposit processing
+
+The Engine interfaces abstract away transaction types with [EIP-2718][eip-2718].
+
+To support rollup-functionality, processing of a new Deposit `TransactionType` is implemented by the engine,
+see [Deposit specification][deposit-spec].
+
+This type of transaction can mint L2 ETH, run EVM, 
+and introduce L1 information to enshrined contracts in the execution state.
+
+> **TODO**: implement deposit spec doc
+
+[deposit-spec]: ./deposits#transaction-type
+
+### Deposit boundaries
+
+Deposits are authenticated by the rollup-node, outside of the engine.
+
+To process deposits safely, the deposits MUST be authenticated first:
+- Ingest directly through trusted Engine API
+- Part of sync towards a trusted block-hash (trusted through previous Engine API instruction)
+
+Deposits MUST never be consumed from the transaction pool.
+*The transaction pool can be disabled in a deposits-only rollup*
+
+
+## Engine API
+
+*Note: the [Engine API][l1-api-spec] is in alpha, `v1.0.0-alpha.5`.
+There may be subtle tweaks, beta starts in a few weeks*
+
+### `engine_forkchoiceUpdatedV1`
+
+This updates which L2 blocks the engine considers to be canonical (`forkchoiceState` argument),
+and optionally initiates block production (`payloadAttributes` argument).
+
+Within the rollup, the types of forkchoice updates translate as:
+- `headBlockHash`: block-hash of the head of the canonical chain. Labeled `"unsafe"` in user JSON-RPC.
+   Nodes may apply L2 blocks out of band ahead of time, and then reorg when L1 data conflicts.
+- `safeBlockHash`: block-hash of the canonical chain, derived from L1 data, unlikely to reorg.
+- `finalizedBlockHash`: irreversible block-hash, matches lower boundary of the dispute-period.
+
+To support rollup-functionality, one backwards-compatible change is introduced
+to [`engine_forkchoiceUpdatedV1`][engine_forkchoiceUpdatedV1]:
+
+[`PayloadAttributesV1`][PayloadAttributesV1] is extended with a `transactions` field,
+equivalent to the `transactions` field in [`ExecutionPayloadV1`][ExecutionPayloadV1]:
+> `Array of DATA` - Array of transaction objects, each object is a byte list (`DATA`) representing
+> `TransactionType || TransactionPayload` or `LegacyTransaction` as defined in [EIP-2718][eip-2718].
+
```

Comment:
There's a chance it will be supported. And as long as it's compatible I don't feel like we have to add a new JSON-RPC method for this.

Reply:
Also regarding JSON-RPC spec: for a very long time ethereum didn't have a good or complete JSON-RPC spec. The execution-apis repo *is* the most canonical spec (user type of API is incomplete, but they do host an interactive site that is generated based on spec docs). The engine API has no other better reference yet. The ethereum github wiki used to host some RPC docs, but those since moved to a separate wiki site (no longer maintained), and then moved to the execution-apis repo. `DATA` and `QUANTITY` are common between all those + lots of EIPs, I'm not sure if the format itself is specified anywhere though.

Reply:
Execution-apis engine document lists it at the top under the "structures" header:
> Fields having DATA and QUANTITY types MUST be encoded according to the HEX value encoding section of Ethereum JSON-RPC API.
>
> Note: Byte order of encoded value having QUANTITY type is big-endian.

I'll add a link.

Reply:
Yes, the link referred to here is the one I used in the rollup node spec.

Did not know executions-apis was also the location for the regular JSON-RPC API, will update the link in the rollup node spec.

Regarding the call, I feel different things with different interface should have different names. I feel like the ETH specs is itself making a big point of that by having `V1` right there in the name. We should preserve that spirit. If it's changed to include the tx field, then we can revert to using the same name.

Reply:
I introduced that very versioning into the Engine API spec (during merge retreat week), if it's compatible it does not need a major versioning change.

Reply:
So you're telling me that if, after the merge, the execution API spec were to change to add a field, the `V1` in the name would not be changed? If that's the case, then fine I suppose.